File: scripts/process-housing-data.py
import csv
import os
import json

# Use current working directory
cwd = os.getcwd()
print(f'Current working directory: {cwd}')

# Try different path approaches
possible_roots = [
    cwd,
    '/vercel/share/v0-project',
    os.path.join(cwd, '..'),
]

project_root = None
for root in possible_roots:
    test_path = os.path.join(root, 'dataset_extraction', 'census_data_output', 'B25002_occupancy_status_zcta.csv')
    if os.path.exists(test_path):
        project_root = root
        print(f'Found project root: {project_root}')
        break

if not project_root:
    print('ERROR: Could not find project root')
    print('Searched in:', possible_roots)
    exit(1)

input_csv = os.path.join(project_root, 'dataset_extraction', 'census_data_output', 'B25002_occupancy_status_zcta.csv')
zipcode_csv = os.path.join(project_root, 'public', 'zipcode-data.csv')
output_csv = os.path.join(project_root, 'public', 'housing-data.csv')

print(f'Reading housing data from: {input_csv}')
print(f'Reading zipcode data from: {zipcode_csv}')

# Read housing data
housing_data = []
with open(input_csv, 'r') as f:
    reader = csv.DictReader(f)
    for row in reader:
        housing_data.append({
            'zcta': row['zcta'],
            'year': int(row['year']),
            'total_units': int(row['total_units']),
            'occupied_units': int(row['occupied_units']),
            'vacant_units': int(row['vacant_units'])
        })

print(f'Loaded {len(housing_data)} housing records')

# Read zipcode coordinates
coordinates_map = {}
with open(zipcode_csv, 'r') as f:
    reader = csv.DictReader(f)
    for row in reader:
        zipcode = row['zipcode']
        coordinates_map[zipcode] = {
            'latitude': float(row['latitude']),
            'longitude': float(row['longitude'])
        }

print(f'Loaded coordinates for {len(coordinates_map)} zipcodes')

# Calculate min and max for normalization
all_total_units = [d['total_units'] for d in housing_data]
min_units = min(all_total_units)
max_units = max(all_total_units)

print(f'Total units range: {min_units} - {max_units}')

# Merge and normalize data
output_data = []
for row in housing_data:
    zipcode = row['zcta']
    if zipcode in coordinates_map:
        # Normalize to 0-100 scale
        normalized_value = ((row['total_units'] - min_units) / (max_units - min_units)) * 100
        
        output_data.append({
            'zipcode': zipcode,
            'year': row['year'],
            'total_units': row['total_units'],
            'occupied_units': row['occupied_units'],
            'vacant_units': row['vacant_units'],
            'normalized_units': round(normalized_value, 2),
            'latitude': coordinates_map[zipcode]['latitude'],
            'longitude': coordinates_map[zipcode]['longitude']
        })

print(f'Created {len(output_data)} records with coordinates')

# Write combined output
with open(output_csv, 'w', newline='') as f:
    fieldnames = ['zipcode', 'year', 'total_units', 'occupied_units', 'vacant_units', 'normalized_units', 'latitude', 'longitude']
    writer = csv.DictWriter(f, fieldnames=fieldnames)
    writer.writeheader()
    writer.writerows(output_data)

print(f'✓ Written processed data to {output_csv}')

# Create year-specific files
for year in range(2011, 2023):
    year_data = [d for d in output_data if d['year'] == year]
    year_path = os.path.join(project_root, 'public', f'housing-data-{year}.csv')
    
    with open(year_path, 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(year_data)
    
    print(f'✓ Created housing-data-{year}.csv with {len(year_data)} records')

print('✓ Data processing complete!')

File: scripts/process-housing-data.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// US ZIP code centroids dataset (a small subset for demonstration)
// In production, you'd want a complete ZIP code database
const zipCodeGeoData = {
  // This is a simplified mapping - you'll need a complete geocoding database
  // For now, we'll use a basic US centroid approach
};

async function fetchZipCodeCoordinates(zipcode) {
  // Using a free geocoding API (Nominatim) with rate limiting
  // In production, consider using a paid service or a local database
  const url = `https://nominatim.openstreetmap.org/search?postalcode=${zipcode}&country=US&format=json`;
  
  try {
    await new Promise(resolve => setTimeout(resolve, 1000)); // Rate limiting
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Housing-Data-Viz/1.0'
      }
    });
    const data = await response.json();
    
    if (data && data.length > 0) {
      return {
        lat: parseFloat(data[0].lat),
        lon: parseFloat(data[0].lon)
      };
    }
  } catch (error) {
    console.error(`Error fetching coordinates for ${zipcode}:`, error.message);
  }
  
  return null;
}

async function processHousingData() {
  console.log('Processing housing data...');
  
  // Read the occupancy status data
  const csvPath = path.join(__dirname, '../dataset_extraction/census_data_output/B25002_occupancy_status_zcta.csv');
  const csvContent = fs.readFileSync(csvPath, 'utf-8');
  
  // Parse CSV
  const lines = csvContent.trim().split('\n');
  const headers = lines[0].split(',');
  
  const data = [];
  const zctaSet = new Set();
  
  // Parse all data
  lines.slice(1).forEach(line => {
    const values = line.split(',');
    const row = {
      zcta: values[0],
      year: parseInt(values[1]),
      total_units: parseInt(values[2]),
      occupied_units: parseInt(values[3]),
      vacant_units: parseInt(values[4])
    };
    data.push(row);
    zctaSet.add(row.zcta);
  });
  
  console.log(`Found ${data.length} records for ${zctaSet.size} unique zip codes`);
  
  // Group by year
  const dataByYear = {};
  for (let year = 2011; year <= 2022; year++) {
    dataByYear[year] = data.filter(d => d.year === year);
  }
  
  // Calculate statistics for normalization
  const allTotalUnits = data.map(d => d.total_units);
  const minUnits = Math.min(...allTotalUnits);
  const maxUnits = Math.max(...allTotalUnits);
  
  console.log(`Total units range: ${minUnits} - ${maxUnits}`);
  
  // For simplicity, we'll use the existing zipcode-data.csv coordinates
  // and match with our housing data
  const existingZipPath = path.join(__dirname, '../public/zipcode-data.csv');
  const existingZipContent = fs.readFileSync(existingZipPath, 'utf-8');
  const existingZipLines = existingZipContent.trim().split('\n');
  
  const coordinatesMap = {};
  existingZipLines.slice(1).forEach(line => {
    const parts = line.split(',');
    if (parts.length >= 5) {
      coordinatesMap[parts[0]] = {
        latitude: parseFloat(parts[3]),
        longitude: parseFloat(parts[4])
      };
    }
  });
  
  console.log(`Loaded coordinates for ${Object.keys(coordinatesMap).length} zip codes`);
  
  // Create output data with coordinates
  const outputData = [];
  
  data.forEach(row => {
    const coords = coordinatesMap[row.zcta];
    if (coords) {
      // Normalize total_units to 0-100 scale
      const normalizedValue = ((row.total_units - minUnits) / (maxUnits - minUnits)) * 100;
      
      outputData.push({
        zipcode: row.zcta,
        year: row.year,
        total_units: row.total_units,
        occupied_units: row.occupied_units,
        vacant_units: row.vacant_units,
        normalized_units: Math.round(normalizedValue * 100) / 100,
        latitude: coords.latitude,
        longitude: coords.longitude
      });
    }
  });
  
  console.log(`Created ${outputData.length} records with coordinates`);
  
  // Write to public directory
  const outputPath = path.join(__dirname, '../public/housing-data.csv');
  
  const csvOutput = [
    'zipcode,year,total_units,occupied_units,vacant_units,normalized_units,latitude,longitude',
    ...outputData.map(d => 
      `${d.zipcode},${d.year},${d.total_units},${d.occupied_units},${d.vacant_units},${d.normalized_units},${d.latitude},${d.longitude}`
    )
  ].join('\n');
  
  fs.writeFileSync(outputPath, csvOutput);
  console.log(`✓ Written processed data to ${outputPath}`);
  
  // Also create year-specific files for easier loading
  for (let year = 2011; year <= 2022; year++) {
    const yearData = outputData.filter(d => d.year === year);
    const yearPath = path.join(__dirname, `../public/housing-data-${year}.csv`);
    
    const yearCsvOutput = [
      'zipcode,year,total_units,occupied_units,vacant_units,normalized_units,latitude,longitude',
      ...yearData.map(d => 
        `${d.zipcode},${d.year},${d.total_units},${d.occupied_units},${d.vacant_units},${d.normalized_units},${d.latitude},${d.longitude}`
      )
    ].join('\n');
    
    fs.writeFileSync(yearPath, yearCsvOutput);
  }
  
  console.log('✓ Created year-specific data files');
  console.log('✓ Data processing complete!');
}

processHousingData().catch(console.error);
